packages/loot-core/migrations/1723665565000_prefs.js

```diff
@@ -0,0 +1,68 @@
+const SYNCED_PREF_KEYS = [
+  'firstDayOfWeekIdx',
+  'dateFormat',
+  'numberFormat',
+  'hideFraction',
+  'isPrivacyEnabled',
+  /^show-extra-balances-/,
+  /^hide-cleared-/,
+  /^parse-date-/,
+  /^csv-mappings-/,
+  /^csv-delimiter-/,
+  /^csv-has-header-/,
+  /^ofx-fallback-missing-payee-/,
+  /^flip-amount-/,
+  // 'budgetType', // TODO: uncomment when `budgetType` moves from metadata to synced prefs
+  /^flags\./,
+];
+
+export default async function runMigration(db, { fs, fileId }) {
+  await db.execQuery(`
+    CREATE TABLE preferences
+       (id TEXT PRIMARY KEY,
+        value TEXT);
+  `);
+
+  try {
+    const budgetDir = fs.getBudgetDir(fileId);
+    const fullpath = fs.join(budgetDir, 'metadata.json');
+
+    const prefs = JSON.parse(await fs.readFile(fullpath));
+
+    if (typeof prefs !== 'object') {
+      return;
+    }
+
+    await Promise.all(
+      Object.keys(prefs).map(async key => {
+        // Check if the current key is of synced-keys type
+        if (
+          !SYNCED_PREF_KEYS.find(keyMatcher =>
+            keyMatcher instanceof RegExp
+              ? keyMatcher.test(key)
+              : keyMatcher === key,
+          )
+        ) {
+          return;
+        }
+
+        // insert the synced prefs in the new table
+        await db.runQuery('INSERT INTO preferences SET id = ?, value = ?', [
+          key,
+          String(prefs[key]),
+        ]);
+
+        // remove the synced prefs from the metadata file
+        delete prefs[key];
+      }),
+    );
+
+    // Update the metadata.json file
+    await fs.writeFile(
+      fs.join(budgetDir, 'metadata.json'),
+      JSON.stringify(prefs),
+    );
+  } catch (e) {
+    // Do nothing
```

Comment:
Just confirming this is OK? 

If migration fails because of file system read/write failure or anything else we just swallow it and pretend successful?

Reply:
I'm getting this when I run this on Electron:

![image](https://github.com/user-attachments/assets/23c6c0d3-2006-49c3-988e-0f698ea76bb7)

Reply:
> Just confirming this is OK?

> If migration fails because of file system read/write failure or anything else we just swallow it and pretend successful?

IMO - as long as the core `preferences` table is created - we're good. Obviously it would be nice to always also migrate all the existing prefs to the new table, but as you pointed out - there's just so many factors to file reads/writes that we can't really guarantee this working 100% of the time. And even if it doesn't work for some cases - that's not the end of the world. Users can still update the prefs afterwards.

---

Not sure I follow the other comment. We already use `String(prefs[key])`. Could you clarify?

Reply:
Apologies, that was a mistake in my console log. 

I'm getting this error when I run it: 

```Server Log: error INSERT INTO preferences SET id = ?, value = ?```

The table is created but no values are inserted:

![image](https://github.com/user-attachments/assets/26ec8f7e-af42-4387-b60c-af21c88b4be3)

Reply:
If you change the insert statement to : 

```
        await db.runQuery('INSERT INTO preferences (id, value) VALUES (?, ?)', [
          key,
          String(prefs[key]),
        ]);
```

it migrates properly

Reply:
Thanks! https://github.com/actualbudget/actual/pull/3423/commits/46b373e6e3c13c358ddd317c88043143eef737e2